{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Mapping Clusters to Specific Features"
     ]
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      "Because of alternative splicing, some features will be overrepresented. To avoid this, features are listed by their type, chromosome, and start & end coordinates. Only unique features from this list are included in subsequent steps. If a cluster intersects a feature whatsoever, that feature will be counted. This means that a cluster spanning a junction will represent multiple features, including the junction itself. \n",
      "This script creates a csv file in the working directory named \"feature_cluster_table.csv\""
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pandas\n",
      "import sys\n",
      "import numpy as np\n",
      "\n",
      "pandas.set_option('line_width', 160)\n",
      "\n",
      "# directory=sys.argv[1] Note again that this has been commented out to allow the script to run in the notebook for this demonstration.\n",
      "directory=\"MiSeqSFPQc.csv_MiSeqNONOc.csvTranscripts/\"   \n",
      "datasets=[]\n",
      "\n",
      "filename_index=pandas.read_csv(directory+\"cluster_and_transcript_annotation_file_index.csv\")\n",
      "translation_status=pandas.read_csv(directory+\"_Translation_Status_Totals.csv\")\n",
      "for x in translation_status.columns[1:]:\n",
      "\tdatasets.append(x)\n",
      "\n",
      "\n",
      "\n",
      "list_of_features=[]\n",
      "def describe_features_within_cluster(cluster):\n",
      "\n",
      "\tcluster_start=cluster['ClusterStart']\n",
      "\tcluster_end=cluster['ClusterEnd']\n",
      "\tmatched_features_list=[]\n",
      "\tfor transcript in eval(cluster['transcripts']):\n",
      "\t\tannotation=pandas.read_csv(directory+transcript+'annotations.csv')\n",
      "\t\t\n",
      "\t\tmatched_features=annotation[annotation['feature_start']<cluster_end]\n",
      "\t\tmatched_features=matched_features[matched_features['feature_end']>cluster_start]\n",
      "\t\tmatched_features['transcript']=transcript\n",
      "\t\tmatched_features_list.append(matched_features)\n",
      "\tif len(matched_features_list)>0:\n",
      "\t\tcluster_features=pandas.concat(matched_features_list)\n",
      "\t\tdel cluster_features['Unnamed: 0']\n",
      "\t\tcluster_features.rename(columns=lambda x: x.replace('feature_classification', 'feature_subtype'), inplace=True)\n",
      "\t\tgroups=cluster_features.groupby(['feature_start','feature_end','feature_subtype'])\n",
      "\t\tgroups=groups.size().reset_index()      #This eliminates replicate features\n",
      "\t\tgroups.columns=[\"Feature Start\",\"Feature End\",\"Feature Subtype\", \"Feature Occurance Count within Cluster\"]\n",
      "\t\t\n",
      "\t\tlist_of_features.extend(groups['Feature Subtype'])\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "feature_dfs=[]\n",
      "for d in datasets:\n",
      "\tlist_of_features=[]\n",
      "\tmatched_reads=pandas.read_csv(directory+d+\"matchedreads.csv\")\n",
      "\tmatched_reads.apply(describe_features_within_cluster,axis=1)\n",
      "\tfeature_subtypes=set(list_of_features)\n",
      "\tprint feature_subtypes\n",
      "\tdf_type_list=[]\n",
      "\tdf_type_count=[]\n",
      "\tfor f in feature_subtypes:\n",
      "\t\tdf_type_list.append(f)\n",
      "\t\tdf_type_count.append(list_of_features.count(f))\n",
      "\tdf_d={\"Feature_Type\":df_type_list,\"%s Count\"%d:df_type_count}\n",
      "\ttotals=pandas.DataFrame(df_d)\n",
      "\ttotals.to_csv(directory+d+\"feature_cluster_table.csv\")\n",
      "\tfeature_dfs.append(totals)\n",
      "\n",
      "\n",
      "t=pandas.merge(left=feature_dfs[0],right=feature_dfs[1],on=\"Feature_Type\",how=\"outer\")\n",
      "t=t.fillna(0)\n",
      "tt=t.T\n",
      "tt.columns=tt.iloc[0,]\n",
      "df = tt.drop(tt.index[0])\n",
      "df=df.div(df.sum(axis=1), axis=0)*100\n",
      "df\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "set([\"Splice Junction Splice Junction Exon from 5'-UTR to 5'-UTR to Intron from 5'-UTR to 5'-UTR to Splice Junction Intron from 5'-UTR to 5'-UTR to Exon from 5'-UTR to Coding_Region\", \"Splice Junction 5'-UTR to 3'-UTR\", \"Exon from 5'-UTR to Coding_Region\", \"Splice Junction Splice Junction Exon from 3'-UTR to 3'-UTR to Intron from 3'-UTR to 3'-UTR to Splice Junction Intron from 3'-UTR to 3'-UTR to Exon from 3'-UTR to 3'-UTR\", 'Intron from Coding_Region to Coding_Region', \"Exon from Coding_Region to 3'-UTR\", \"Splice Junction Splice Junction Exon from Coding_Region to Coding_Region to Intron from Coding_Region to Coding_Region to Splice Junction Intron from Coding_Region to Coding_Region to Exon from Coding_Region to 3'-UTR\", \"Splice Junction Splice Junction Exon from Coding_Region to 3'-UTR to Intron from 3'-UTR to 3'-UTR to Splice Junction Intron from 3'-UTR to 3'-UTR to Exon from 3'-UTR to 3'-UTR\", 'Exon in a ncRNA', 'Splice Junction Splice Junction Exon in a ncRNA to Intron in a ncRNA to Splice Junction Intron in a ncRNA to Exon in a ncRNA', \"Splice Junction Splice Junction Exon from 5'-UTR to Coding_Region to Intron from Coding_Region to Coding_Region to Splice Junction Intron from Coding_Region to Coding_Region to Exon from Coding_Region to Coding_Region\", \"Exon from 5'-UTR to 3'-UTR\", \"Intron from 3'-UTR to 3'-UTR\", 'Intron in a ncRNA', \"3'-UTR\", \"Intron from 5'-UTR to 5'-UTR\", \"5'-UTR\", \"Exon from 3'-UTR to 3'-UTR\", 'Splice Junction Splice Junction Exon from Coding_Region to Coding_Region to Intron from Coding_Region to Coding_Region to Splice Junction Intron from Coding_Region to Coding_Region to Exon from Coding_Region to Coding_Region', \"Splice Junction Splice Junction Exon from 5'-UTR to 5'-UTR to Intron from 5'-UTR to 5'-UTR to Splice Junction Intron from 5'-UTR to 5'-UTR to Exon from 5'-UTR to 5'-UTR\"])\n",
        "set([\"Splice Junction Splice Junction Exon from 5'-UTR to 5'-UTR to Intron from 5'-UTR to 5'-UTR to Splice Junction Intron from 5'-UTR to 5'-UTR to Exon from 5'-UTR to Coding_Region\", \"Exon from 5'-UTR to 5'-UTR\", \"Splice Junction Splice Junction Exon from 5'-UTR to 5'-UTR to Intron from 5'-UTR to 5'-UTR to Splice Junction Intron from 5'-UTR to 5'-UTR to Exon from 5'-UTR to 5'-UTR\", \"Exon from 5'-UTR to Coding_Region\", 'Intron from Coding_Region to Coding_Region', \"Splice Junction Splice Junction Exon from Coding_Region to Coding_Region to Intron from Coding_Region to Coding_Region to Splice Junction Intron from Coding_Region to Coding_Region to Exon from Coding_Region to 3'-UTR\", \"Splice Junction Splice Junction Exon from Coding_Region to 3'-UTR to Intron from 3'-UTR to 3'-UTR to Splice Junction Intron from 3'-UTR to 3'-UTR to Exon from 3'-UTR to 3'-UTR\", 'Exon in a ncRNA', \"Splice Junction Splice Junction Exon from 5'-UTR to Coding_Region to Intron from Coding_Region to Coding_Region to Splice Junction Intron from Coding_Region to Coding_Region to Exon from Coding_Region to 3'-UTR\", 'Splice Junction Splice Junction Exon in a ncRNA to Intron in a ncRNA to Splice Junction Intron in a ncRNA to Exon in a ncRNA', \"Splice Junction Splice Junction Exon from 5'-UTR to Coding_Region to Intron from Coding_Region to Coding_Region to Splice Junction Intron from Coding_Region to Coding_Region to Exon from Coding_Region to Coding_Region\", \"Exon from 5'-UTR to 3'-UTR\", \"Intron from 3'-UTR to 3'-UTR\", 'Intron in a ncRNA', \"3'-UTR\", \"Intron from 5'-UTR to 5'-UTR\", \"5'-UTR\", 'Splice Junction Splice Junction Exon from Coding_Region to Coding_Region to Intron from Coding_Region to Coding_Region to Splice Junction Intron from Coding_Region to Coding_Region to Exon from Coding_Region to Coding_Region', \"Splice Junction Splice Junction Exon from 3'-UTR to 3'-UTR to Intron from 3'-UTR to 3'-UTR to Splice Junction Intron from 3'-UTR to 3'-UTR to Exon from 3'-UTR to 3'-UTR\"])"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n"
       ]
      },
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th>Feature_Type</th>\n",
        "      <th>Splice Junction Splice Junction Exon from 5'-UTR to 5'-UTR to Intron from 5'-UTR to 5'-UTR to Splice Junction Intron from 5'-UTR to 5'-UTR to Exon from 5'-UTR to Coding_Region</th>\n",
        "      <th>Splice Junction 5'-UTR to 3'-UTR</th>\n",
        "      <th>Exon from 5'-UTR to Coding_Region</th>\n",
        "      <th>Splice Junction Splice Junction Exon from 3'-UTR to 3'-UTR to Intron from 3'-UTR to 3'-UTR to Splice Junction Intron from 3'-UTR to 3'-UTR to Exon from 3'-UTR to 3'-UTR</th>\n",
        "      <th>Intron from Coding_Region to Coding_Region</th>\n",
        "      <th>Exon from Coding_Region to 3'-UTR</th>\n",
        "      <th>Splice Junction Splice Junction Exon from Coding_Region to Coding_Region to Intron from Coding_Region to Coding_Region to Splice Junction Intron from Coding_Region to Coding_Region to Exon from Coding_Region to 3'-UTR</th>\n",
        "      <th>Splice Junction Splice Junction Exon from Coding_Region to 3'-UTR to Intron from 3'-UTR to 3'-UTR to Splice Junction Intron from 3'-UTR to 3'-UTR to Exon from 3'-UTR to 3'-UTR</th>\n",
        "      <th>Exon in a ncRNA</th>\n",
        "      <th>Splice Junction Splice Junction Exon in a ncRNA to Intron in a ncRNA to Splice Junction Intron in a ncRNA to Exon in a ncRNA</th>\n",
        "      <th>Splice Junction Splice Junction Exon from 5'-UTR to Coding_Region to Intron from Coding_Region to Coding_Region to Splice Junction Intron from Coding_Region to Coding_Region to Exon from Coding_Region to Coding_Region</th>\n",
        "      <th>Exon from 5'-UTR to 3'-UTR</th>\n",
        "      <th>Intron from 3'-UTR to 3'-UTR</th>\n",
        "      <th>Intron in a ncRNA</th>\n",
        "      <th>3'-UTR</th>\n",
        "      <th>Intron from 5'-UTR to 5'-UTR</th>\n",
        "      <th>5'-UTR</th>\n",
        "      <th>Exon from 3'-UTR to 3'-UTR</th>\n",
        "      <th>Splice Junction Splice Junction Exon from Coding_Region to Coding_Region to Intron from Coding_Region to Coding_Region to Splice Junction Intron from Coding_Region to Coding_Region to Exon from Coding_Region to Coding_Region</th>\n",
        "      <th>Splice Junction Splice Junction Exon from 5'-UTR to 5'-UTR to Intron from 5'-UTR to 5'-UTR to Splice Junction Intron from 5'-UTR to 5'-UTR to Exon from 5'-UTR to 5'-UTR</th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>MiSeqSFPQc.csv Count</th>\n",
        "      <td> 2.477701</td>\n",
        "      <td> 0.09910803</td>\n",
        "      <td> 0.2973241</td>\n",
        "      <td>  1.288404</td>\n",
        "      <td> 15.65907</td>\n",
        "      <td> 0.1982161</td>\n",
        "      <td> 1.387512</td>\n",
        "      <td> 0.8919722</td>\n",
        "      <td> 16.74926</td>\n",
        "      <td> 0.6937562</td>\n",
        "      <td> 4.757185</td>\n",
        "      <td> 2.378593</td>\n",
        "      <td> 6.640238</td>\n",
        "      <td> 9.018831</td>\n",
        "      <td> 9.712587</td>\n",
        "      <td> 9.811695</td>\n",
        "      <td> 9.712587</td>\n",
        "      <td> 0.09910803</td>\n",
        "      <td> 5.153617</td>\n",
        "      <td> 2.973241</td>\n",
        "      <td>...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>MiSeqNONOc.csv Count</th>\n",
        "      <td>  1.62037</td>\n",
        "      <td>          0</td>\n",
        "      <td> 0.2314815</td>\n",
        "      <td> 0.6944444</td>\n",
        "      <td> 14.12037</td>\n",
        "      <td>         0</td>\n",
        "      <td> 1.388889</td>\n",
        "      <td>  1.157407</td>\n",
        "      <td> 17.36111</td>\n",
        "      <td> 0.3472222</td>\n",
        "      <td> 3.587963</td>\n",
        "      <td> 4.166667</td>\n",
        "      <td> 7.060185</td>\n",
        "      <td> 11.34259</td>\n",
        "      <td> 11.11111</td>\n",
        "      <td> 9.722222</td>\n",
        "      <td> 9.722222</td>\n",
        "      <td>          0</td>\n",
        "      <td> 4.282407</td>\n",
        "      <td> 1.851852</td>\n",
        "      <td>...</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>2 rows \u00d7 22 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 26,
       "text": [
        "Feature_Type         Splice Junction Splice Junction Exon from 5'-UTR to 5'-UTR to Intron from 5'-UTR to 5'-UTR to Splice Junction Intron from 5'-UTR to 5'-UTR to Exon from 5'-UTR to Coding_Region  \\\n",
        "MiSeqSFPQc.csv Count                                           2.477701                                                                                                                                \n",
        "MiSeqNONOc.csv Count                                            1.62037                                                                                                                                \n",
        "\n",
        "Feature_Type         Splice Junction 5'-UTR to 3'-UTR Exon from 5'-UTR to Coding_Region  \\\n",
        "MiSeqSFPQc.csv Count                       0.09910803                         0.2973241   \n",
        "MiSeqNONOc.csv Count                                0                         0.2314815   \n",
        "\n",
        "Feature_Type         Splice Junction Splice Junction Exon from 3'-UTR to 3'-UTR to Intron from 3'-UTR to 3'-UTR to Splice Junction Intron from 3'-UTR to 3'-UTR to Exon from 3'-UTR to 3'-UTR  \\\n",
        "MiSeqSFPQc.csv Count                                           1.288404                                                                                                                         \n",
        "MiSeqNONOc.csv Count                                          0.6944444                                                                                                                         \n",
        "\n",
        "Feature_Type         Intron from Coding_Region to Coding_Region Exon from Coding_Region to 3'-UTR  \\\n",
        "MiSeqSFPQc.csv Count                                   15.65907                         0.1982161   \n",
        "MiSeqNONOc.csv Count                                   14.12037                                 0   \n",
        "\n",
        "Feature_Type         Splice Junction Splice Junction Exon from Coding_Region to Coding_Region to Intron from Coding_Region to Coding_Region to Splice Junction Intron from Coding_Region to Coding_Region to Exon from Coding_Region to 3'-UTR  \\\n",
        "MiSeqSFPQc.csv Count                                           1.387512                                                                                                                                                                          \n",
        "MiSeqNONOc.csv Count                                           1.388889                                                                                                                                                                          \n",
        "\n",
        "Feature_Type         Splice Junction Splice Junction Exon from Coding_Region to 3'-UTR to Intron from 3'-UTR to 3'-UTR to Splice Junction Intron from 3'-UTR to 3'-UTR to Exon from 3'-UTR to 3'-UTR  \\\n",
        "MiSeqSFPQc.csv Count                                          0.8919722                                                                                                                                \n",
        "MiSeqNONOc.csv Count                                           1.157407                                                                                                                                \n",
        "\n",
        "Feature_Type         Exon in a ncRNA  \\\n",
        "MiSeqSFPQc.csv Count        16.74926   \n",
        "MiSeqNONOc.csv Count        17.36111   \n",
        "\n",
        "Feature_Type         Splice Junction Splice Junction Exon in a ncRNA to Intron in a ncRNA to Splice Junction Intron in a ncRNA to Exon in a ncRNA  \\\n",
        "MiSeqSFPQc.csv Count                                          0.6937562                                                                             \n",
        "MiSeqNONOc.csv Count                                          0.3472222                                                                             \n",
        "\n",
        "Feature_Type         Splice Junction Splice Junction Exon from 5'-UTR to Coding_Region to Intron from Coding_Region to Coding_Region to Splice Junction Intron from Coding_Region to Coding_Region to Exon from Coding_Region to Coding_Region  \\\n",
        "MiSeqSFPQc.csv Count                                           4.757185                                                                                                                                                                          \n",
        "MiSeqNONOc.csv Count                                           3.587963                                                                                                                                                                          \n",
        "\n",
        "Feature_Type         Exon from 5'-UTR to 3'-UTR Intron from 3'-UTR to 3'-UTR Intron in a ncRNA    3'-UTR Intron from 5'-UTR to 5'-UTR    5'-UTR  \\\n",
        "MiSeqSFPQc.csv Count                   2.378593                     6.640238          9.018831  9.712587                     9.811695  9.712587   \n",
        "MiSeqNONOc.csv Count                   4.166667                     7.060185          11.34259  11.11111                     9.722222  9.722222   \n",
        "\n",
        "Feature_Type         Exon from 3'-UTR to 3'-UTR  \\\n",
        "MiSeqSFPQc.csv Count                 0.09910803   \n",
        "MiSeqNONOc.csv Count                          0   \n",
        "\n",
        "Feature_Type         Splice Junction Splice Junction Exon from Coding_Region to Coding_Region to Intron from Coding_Region to Coding_Region to Splice Junction Intron from Coding_Region to Coding_Region to Exon from Coding_Region to Coding_Region  \\\n",
        "MiSeqSFPQc.csv Count                                           5.153617                                                                                                                                                                                 \n",
        "MiSeqNONOc.csv Count                                           4.282407                                                                                                                                                                                 \n",
        "\n",
        "Feature_Type         Splice Junction Splice Junction Exon from 5'-UTR to 5'-UTR to Intron from 5'-UTR to 5'-UTR to Splice Junction Intron from 5'-UTR to 5'-UTR to Exon from 5'-UTR to 5'-UTR  \\\n",
        "MiSeqSFPQc.csv Count                                           2.973241                                                                                                                         \n",
        "MiSeqNONOc.csv Count                                           1.851852                                                                                                                         \n",
        "\n",
        "Feature_Type              \n",
        "MiSeqSFPQc.csv Count ...  \n",
        "MiSeqNONOc.csv Count ...  \n",
        "\n",
        "[2 rows x 22 columns]"
       ]
      }
     ],
     "prompt_number": 26
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df2=df\n",
      "try:\n",
      "    df2['mRNA_CDS_exons']=df[\"Exon from Coding_Region to Coding_Region\"]\n",
      "except:\n",
      "    pass\n",
      "\n",
      "df2['mRNA_UTR_exons']=df[\"Exon from 5'-UTR to Coding_Region\"]+df[\"Exon from 3'-UTR to 3'-UTR\"]+df[\"Exon from 5'-UTR to 5'-UTR\"]+df[\"Exon from 5'-UTR to 3'-UTR\"]+df[\"Exon from Coding_Region to 3'-UTR\"]\n",
      "df2['ncRNA_exons']=df[\"Exon in a ncRNA\"]\n",
      "df2[\"3'-UTR\"]=df[\"3'-UTR\"]\n",
      "df2[\"5'-UTR\"]=df[\"5'-UTR\"]\n",
      "df2[\"mRNA_Introns\"]=df[\"Intron from Coding_Region to Coding_Region\"]+df[\"Intron from 3'-UTR to 3'-UTR\"]+df[\"Intron from 5'-UTR to 5'-UTR\"]\n",
      "df2[\"ncRNA_Introns\"]=df[\"Intron in a ncRNA\"]\n",
      "cols=[x for x in df.columns.values.tolist() if \"Junction\" in x]\n",
      "df2[\"Splicing Junction\"]=df[cols].sum(axis=1)\n",
      "try:\n",
      "    df2=df2[['mRNA_CDS_exons','mRNA_UTR_exons','ncRNA_exons','UTR3','UTR5',\"mRNA_Introns\",\"Splicing Junction\"]]\n",
      "except:\n",
      "    df2=df2[['mRNA_UTR_exons','ncRNA_exons',\"3'-UTR\",\"5'-UTR\",\"mRNA_Introns\",\"ncRNA_Introns\",\"Splicing Junction\"]]\n",
      "df2\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>mRNA_UTR_exons</th>\n",
        "      <th>ncRNA_exons</th>\n",
        "      <th>3'-UTR</th>\n",
        "      <th>5'-UTR</th>\n",
        "      <th>mRNA_Introns</th>\n",
        "      <th>ncRNA_Introns</th>\n",
        "      <th>Splicing Junction</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>MiSeqSFPQc.csv Count</th>\n",
        "      <td> 2.973241</td>\n",
        "      <td> 16.74926</td>\n",
        "      <td> 9.712587</td>\n",
        "      <td> 9.712587</td>\n",
        "      <td>   32.111</td>\n",
        "      <td> 9.018831</td>\n",
        "      <td> 19.722498</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>MiSeqNONOc.csv Count</th>\n",
        "      <td> 4.513889</td>\n",
        "      <td> 17.36111</td>\n",
        "      <td> 11.11111</td>\n",
        "      <td> 9.722222</td>\n",
        "      <td> 30.90278</td>\n",
        "      <td> 11.34259</td>\n",
        "      <td> 15.046296</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>2 rows \u00d7 7 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 51,
       "text": [
        "                     mRNA_UTR_exons ncRNA_exons    3'-UTR    5'-UTR mRNA_Introns ncRNA_Introns  Splicing Junction\n",
        "MiSeqSFPQc.csv Count       2.973241    16.74926  9.712587  9.712587       32.111      9.018831          19.722498\n",
        "MiSeqNONOc.csv Count       4.513889    17.36111  11.11111  9.722222     30.90278      11.34259          15.046296\n",
        "\n",
        "[2 rows x 7 columns]"
       ]
      }
     ],
     "prompt_number": 51
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import vincent \n",
      "vincent.core.initialize_notebook()\n",
      "stack = vincent.StackedBar(df2)\n",
      "stack.legend(title=\"Feature Type\")\n",
      "stack.colors(brew='Set3')\n",
      "stack.axis_titles(x=\"Sample\",y=\"Proportions of Bound Features, by %\")\n",
      "stack.display()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "\n",
        "           <script>\n",
        "               \n",
        "                function vct_load_lib(url, callback){\n",
        "                      if(typeof d3 !== 'undefined' &&\n",
        "                         url === 'http://d3js.org/d3.v3.min.js'){\n",
        "                        callback()\n",
        "                      }\n",
        "                      var s = document.createElement('script');\n",
        "                      s.src = url;\n",
        "                      s.async = true;\n",
        "                      s.onreadystatechange = s.onload = callback;\n",
        "                      s.onerror = function(){\n",
        "                        console.warn(\"failed to load library \" + url);\n",
        "                        };\n",
        "                      document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
        "                };\n",
        "                var vincent_event = new CustomEvent(\n",
        "                  \"vincent_libs_loaded\",\n",
        "                  {bubbles: true, cancelable: true}\n",
        "                );\n",
        "                \n",
        "               function load_all_libs(){\n",
        "                  console.log('Loading Vincent libs...')\n",
        "                  vct_load_lib('http://d3js.org/d3.v3.min.js', function(){\n",
        "                  vct_load_lib('http://d3js.org/d3.geo.projection.v0.min.js', function(){\n",
        "                  vct_load_lib('http://wrobstory.github.io/d3-cloud/d3.layout.cloud.js', function(){\n",
        "                  vct_load_lib('http://wrobstory.github.io/vega/vega.v1.3.3.js', function(){\n",
        "                  window.dispatchEvent(vincent_event);\n",
        "                  });\n",
        "                  });\n",
        "                  });\n",
        "                  });\n",
        "               };\n",
        "               if(typeof define === \"function\" && define.amd){\n",
        "                    if (window['d3'] === undefined ||\n",
        "                        window['topojson'] === undefined){\n",
        "                        require.config(\n",
        "                            {paths: {\n",
        "                              d3: 'http://d3js.org/d3.v3.min',\n",
        "                              topojson: 'http://d3js.org/topojson.v1.min'\n",
        "                              }\n",
        "                            }\n",
        "                          );\n",
        "                        require([\"d3\"], function(d3){\n",
        "                            console.log('Loading Vincent from require.js...')\n",
        "                            window.d3 = d3;\n",
        "                            require([\"topojson\"], function(topojson){\n",
        "                                window.topojson = topojson;\n",
        "                                load_all_libs();\n",
        "                            });\n",
        "                        });\n",
        "                    } else {\n",
        "                        load_all_libs();\n",
        "                    };\n",
        "               }else{\n",
        "                    console.log('Require.js not found, loading manually...')\n",
        "                    load_all_libs();\n",
        "               };\n",
        "\n",
        "           </script>"
       ],
       "metadata": {},
       "output_type": "display_data",
       "text": [
        "<IPython.core.display.HTML at 0x1073e78d0>"
       ]
      },
      {
       "html": [
        "<div id=\"vis1139140c7603487aa2f0aee8da4dac1b\"></div>\n",
        "<script>\n",
        "   ( function() {\n",
        "     var _do_plot = function() {\n",
        "       if (typeof vg === 'undefined') {\n",
        "         window.addEventListener('vincent_libs_loaded', _do_plot)\n",
        "         return;\n",
        "       }\n",
        "       vg.parse.spec({\"axes\": [{\"scale\": \"x\", \"title\": \"Sample\", \"type\": \"x\"}, {\"scale\": \"y\", \"title\": \"Proportions of Bound Features, by %\", \"type\": \"y\"}], \"data\": [{\"name\": \"table\", \"values\": [{\"col\": \"mRNA_UTR_exons\", \"idx\": \"MiSeqSFPQc.csv Count\", \"val\": 2.9732408325074333}, {\"col\": \"ncRNA_exons\", \"idx\": \"MiSeqSFPQc.csv Count\", \"val\": 16.74925668979187}, {\"col\": \"3'-UTR\", \"idx\": \"MiSeqSFPQc.csv Count\", \"val\": 9.712586719524282}, {\"col\": \"5'-UTR\", \"idx\": \"MiSeqSFPQc.csv Count\", \"val\": 9.712586719524282}, {\"col\": \"mRNA_Introns\", \"idx\": \"MiSeqSFPQc.csv Count\", \"val\": 32.111000991080275}, {\"col\": \"ncRNA_Introns\", \"idx\": \"MiSeqSFPQc.csv Count\", \"val\": 9.018830525272548}, {\"col\": \"Splicing Junction\", \"idx\": \"MiSeqSFPQc.csv Count\", \"val\": 19.722497522299303}, {\"col\": \"mRNA_UTR_exons\", \"idx\": \"MiSeqNONOc.csv Count\", \"val\": 4.513888888888888}, {\"col\": \"ncRNA_exons\", \"idx\": \"MiSeqNONOc.csv Count\", \"val\": 17.36111111111111}, {\"col\": \"3'-UTR\", \"idx\": \"MiSeqNONOc.csv Count\", \"val\": 11.11111111111111}, {\"col\": \"5'-UTR\", \"idx\": \"MiSeqNONOc.csv Count\", \"val\": 9.722222222222223}, {\"col\": \"mRNA_Introns\", \"idx\": \"MiSeqNONOc.csv Count\", \"val\": 30.90277777777778}, {\"col\": \"ncRNA_Introns\", \"idx\": \"MiSeqNONOc.csv Count\", \"val\": 11.342592592592593}, {\"col\": \"Splicing Junction\", \"idx\": \"MiSeqNONOc.csv Count\", \"val\": 15.046296296296294}]}, {\"name\": \"stats\", \"source\": \"table\", \"transform\": [{\"keys\": [\"data.idx\"], \"type\": \"facet\"}, {\"type\": \"stats\", \"value\": \"data.val\"}]}], \"height\": 500, \"legends\": [{\"fill\": \"color\", \"offset\": 0, \"properties\": {}, \"title\": \"Feature Type\"}], \"marks\": [{\"from\": {\"data\": \"table\", \"transform\": [{\"keys\": [\"data.col\"], \"type\": \"facet\"}, {\"height\": \"data.val\", \"point\": \"data.idx\", \"type\": \"stack\"}]}, \"marks\": [{\"properties\": {\"enter\": {\"fill\": {\"field\": \"data.col\", \"scale\": \"color\"}, \"width\": {\"band\": true, \"offset\": -1, \"scale\": \"x\"}, \"x\": {\"field\": \"data.idx\", \"scale\": \"x\"}, \"y\": {\"field\": \"y\", \"scale\": \"y\"}, \"y2\": {\"field\": \"y2\", \"scale\": \"y\"}}}, \"type\": \"rect\"}], \"type\": \"group\"}], \"padding\": \"auto\", \"scales\": [{\"domain\": {\"data\": \"table\", \"field\": \"data.idx\"}, \"name\": \"x\", \"range\": \"width\", \"type\": \"ordinal\", \"zero\": false}, {\"domain\": {\"data\": \"stats\", \"field\": \"sum\"}, \"name\": \"y\", \"nice\": true, \"range\": \"height\"}, {\"domain\": {\"data\": \"table\", \"field\": \"data.col\"}, \"name\": \"color\", \"range\": [\"#8dd3c7\", \"#ffffb3\", \"#bebada\", \"#fb8072\", \"#80b1d3\", \"#fdb462\", \"#b3de69\", \"#fccde5\", \"#d9d9d9\", \"#bc80bd\", \"#ccebc5\", \"#ffed6f\"], \"type\": \"ordinal\"}], \"width\": 960}, function(chart) {\n",
        "         chart({el: \"#vis1139140c7603487aa2f0aee8da4dac1b\"}).update();\n",
        "       });\n",
        "     };\n",
        "     _do_plot();\n",
        "   })();\n",
        "</script>\n",
        "<style>.vega canvas {width: 100%;}</style>\n",
        "        "
       ],
       "metadata": {},
       "output_type": "display_data",
       "text": [
        "<IPython.core.display.HTML at 0x1073e7590>"
       ]
      }
     ],
     "prompt_number": 52
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}